File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/semaphores/Singleton.cs
/* Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com */
using System;
using com.db4o;
using com.db4o.query;
using j4o.lang;

namespace com.db4odoc.f1.semaphores
{
	/**
	 * This class	demonstrates the use	of a	semaphore to ensure that only
	 * one instance of a	certain	class is stored to an ObjectContainer.
	 * 
	 * Caution	!!! The getSingleton method contains a commit()	call.	
	 */
	public class	Singleton 
	{
		/**
		 * returns	a singleton object of one	class for an	ObjectContainer.
		 * <br><b>Caution !!!	This	method contains a commit() call.</b> 
		 */
		public static	Object GetSingleton(ObjectContainer objectContainer, Class	clazz) 
		{
			Object obj =	queryForSingletonClass(objectContainer, clazz);
			if (obj != null) 
			{
				return obj;
			}

			String semaphore =	"Singleton#getSingleton_" + clazz.GetName();

			if (!objectContainer.Ext().SetSemaphore(semaphore,	10000)) 
			{
				throw new Exception("Blocked semaphore "	+ semaphore);
			}

			obj =	queryForSingletonClass(objectContainer, clazz);

			if (obj == null) 
			{

				try 
				{
					obj =	clazz.NewInstance();
				} 
				catch (Exception e) 
				{
					System.Console.WriteLine(e.Message);
				} 

				objectContainer.Set(obj);

				/* !!! CAUTION !!!
				 * There is a	commit	call here.
				 * 
				 * The commit call	is	necessary, so	other transactions
				 * can see the new inserted object.
				 */
				objectContainer.Commit();

			}

			objectContainer.Ext().ReleaseSemaphore(semaphore);

			return obj;
		}

		private	static Object queryForSingletonClass(ObjectContainer objectContainer, Class	clazz) 
		{
			Query q = objectContainer.Query();
			q.Constrain(clazz);
			ObjectSet	objectSet	= q.Execute();
			if (objectSet.Size() == 1)	
			{
				return objectSet.Next();
			}
			if (objectSet.Size() > 1) 
			{
				throw new Exception(
					"Singleton problem. Multiple	instances of: "	+ clazz.GetName());
			}
			return null;
		}

	}

}

File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/semaphores/Singleton.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 
Imports System
Imports com.db4o
Imports com.db4o.query

Namespace com.db4odoc.f1.semaphores
    ' This class	demonstrates the use	of a	semaphore to ensure that only
    ' one instance of a	certain	class is stored to an ObjectContainer.
    ' 
    ' Caution	!!! The getSingleton method contains a commit()	call.	

    Public Class Singleton

        ' returns	a singleton object of one	class for an	ObjectContainer.
        ' Caution !!!	This	method contains a commit() call.

        Public Shared Function GetSingleton(ByVal objectContainer As ObjectContainer, ByVal clazz As j4o.lang.Class) As Object
            Dim obj As Object = queryForSingletonClass(objectContainer, clazz)
            If Not obj Is Nothing Then
                Return obj
            End If

            Dim semaphore As String = "Singleton#getSingleton_" + clazz.GetName()

            If Not objectContainer.Ext().SetSemaphore(semaphore, 10000) Then
                Throw New Exception("Blocked semaphore " + semaphore)
            End If

            obj = queryForSingletonClass(objectContainer, clazz)

            If obj Is Nothing Then

                Try
                    obj = clazz.NewInstance()
                Catch e As Exception
                    System.Console.WriteLine(e.Message)
                End Try

                objectContainer.Set(obj)
                ' 				 Not Not Not  CAUTION Not Not Not 
                ' 				 * There is a	commit	call here.
                ' 				 * 
                ' 				 * The commit call	is	necessary, so	other transactions
                ' 				 * can see the New inserted Object.
                ' 				 */
                objectContainer.Commit()

            End If

            objectContainer.Ext().ReleaseSemaphore(semaphore)

            Return obj
        End Function

        Private Shared Function queryForSingletonClass(ByVal objectContainer As ObjectContainer, ByVal clazz As j4o.lang.Class) As Object
            Dim q As Query = objectContainer.Query()
            q.Constrain(clazz)
            Dim objectSet As ObjectSet = q.Execute()
            If objectSet.Size() = 1 Then
                Return objectSet.Next()
            End If
            If objectSet.Size() > 1 Then
                Throw New Exception("Singleton problem. Multiple	instances of: " + clazz.GetName())
            End If
            Return Nothing
        End Function

    End Class

End Namespace
